Modeling HVAC Operational Faults – Chiller Supply Water Temperature Sensor Offset

rongpengzhang opened the pull request:
## Pull request overview

The fault modeling features aim to model the HVAC operational faults to improve the accuracy of building simulations and to support the building retrofitting. Compared with Phase I and II that focused on the faults related with air loop, Phase III will include the faults in the plant loop. Four new types of fault will be implemented in Phase III, namely:
1.  Chiller Supply Water Temperature Sensor Offset
2.  Condenser Supply Water Temperature Sensor Offset
3.  Cooling Tower Scaling
4.  Coil Supply Air Temperature Sensor Offset 

This branch handles the 1st fault in the list. It also restructures the existing faults models by introducing  more OO features to the FaultsManager routine.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [x] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [x] If transition, add rules to spreadsheet
  - [x] If transition, add transition source
  - [x] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [x] Documentation changes in place
- [x] ExpandObjects changes??
- [x] If output changes, including tabular output structure, add to output rules file for interfaces

Compared with Phase I and II that focused on the faults related with air
loop, Phase III will include the faults in the plant loop. Four new
types of fault will be implemented in Phase III, namely:
1. Chiller Supply Water Temperature Sensor Offset
2. Condenser Supply Water Temperature Sensor Offset
3. Cooling Tower Scaling
4. Coil Supply Air Temperature Sensor Offset

rongpengzhang committed:
NFP for four fault models

Compared with Phase I and II that focused on the faults related with air
loop, Phase III will include the faults in the plant loop. Four new
types of fault will be implemented in Phase III, namely:

1. Chiller Supply Water Temperature Sensor Offset
2. Condenser Supply Water Temperature Sensor Offset
3. Cooling Tower Scaling
4. Coil Supply Air Temperature Sensor Offset

rongpengzhang committed:
V1. Fault Manager update to read in all the four new fault models

rongpengzhang committed:
V2. Fault Manager restructure for better code organization of new fault types

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang committed:
V3.1: OO refactor for FaultModel:Fouling:AirFilter

rongpengzhang committed:
V3.2: OO refactor for other fault models

rongpengzhang committed:
V3.3: OO refactor minor bug fix

rongpengzhang committed:
V4. IDD design for the FaultyChillerSWTSensor

rongpengzhang committed:
V5. Fault Struct initialization for FaultyChillerSWTSensor

rongpengzhang committed:
V6. Apply fault to Chiller:Electric:ReformulatedEIR

rongpengzhang committed:
Member function to calculate actual offset

rongpengzhang committed:
Member function to calculate chiller with SWT fault

rongpengzhang committed:
Apply the fault to Chiller:Electric:ReformulatedEIR

rongpengzhang committed:
DD for four fault models

rongpengzhang committed:
Format refinement of DD

rongpengzhang committed:
Merge branch '120694307_Fault_ChillerSWTSensorOffset' of https://github.com/NREL/EnergyPlus into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang committed:
Minor updates of DD

rongpengzhang committed:
V7. Apply the fault to Chiller:Electric:EIR

rongpengzhang committed:
V8.1 Apply the fault to Chiller:Electric

rongpengzhang committed:
V8.2 Apply the fault to Chiller:EngineDriven

rongpengzhang committed:
V8.3 Apply the fault to Chiller:CombustionTurbine

rongpengzhang committed:
V8.4 Apply the fault to Chiller:ConstantCOP

rongpengzhang committed:
V9 Apply the fault to Chiller:Absorption & Chiller:Absorption:Indirect

rongpengzhang committed:
Unit Test

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -79051,6 +79051,43 @@ FaultModel:EnthalpySensorOffset:ReturnAir,
+   
+FaultModel:TemperatureSensorOffset:ChillerSupplyWater,
+   \memo This object describes fault of chiller supply water temperature sensor offset
+   \min-fields 3
```
min-fields 6 (should include every field that has a default).

rongpengzhang replied:
Addressed.

mjwitte commented on src/EnergyPlus/FaultsManager.cc:
```diff
@@ -218,355 +277,747 @@ namespace FaultsManager {
+		// SUBROUTINE INFORMATION:
+		//       AUTHOR         Rongpeng Zhang
+		//       DATE WRITTEN   Jun. 2016
+		//       MODIFIED       na
```
Minor detail - these na comments and other unused comment blocks below can be omitted.  The old comment template does not apply to new code.

rongpengzhang replied:
@mjwitte I like the old comment template which is very organized, but anyway, I've followed the new rules and cleaned up the unused comments for all the functions in the FaultsManager routine.

rongpengzhang committed:
Example file

rongpengzhang committed:
Documentation

rongpengzhang committed:
Address comments

rongpengzhang committed:
minor

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

mjwitte commented on doc/input-output-reference/src/overview/group-operational-faults.tex:
```diff
@@ -353,9 +363,11 @@ \subsubsection{Inputs}\label{inputs-7-015}
 This field provides the name of a Thermostat Offset Fault object that causes the humidistat offset fault. It should be the name of a FaultModel:ThermostatOffset object. This field is required for the Type ThermostatOffsetDependent. This is used as a multiplier to the reference humidistat offset value. This schedule should be set to a non-zero value when a fault is applicable and ``0.0'' when it i ...
 
-\subsection{An example of IDF codes for the thermostat/humidistat offset fault models:}\label{an-example-of-idf-codes-for-the-thermostathumidistat-offset-fault-models}
+\paragraph{IDF Examples}\label{an-example-of-idf-codes-for-the-thermostathumidistat-offset-fault-models}
```
The idf example that come prior to FaultModel:ThermostatOffset are misaligned.  Would be good to clean them up here.  All of the other fault model idf examples look good.

rongpengzhang replied:
Addressed.

mjwitte commented on doc/engineering-reference/src/operational-faults.tex:
```diff
@@ -16,7 +16,10 @@ \subsection{Introduction to Operational Faults Modeling}\label{introduction-to-o
+EnergyPlus contains a number of objects to model operational faults of sensors, meters, equipment and systems. The current implementation allows the modeling of the following fault types: (1) sensor faults with air economizers, (2) thermostat/humidistat offset faults, (3) fouling or scaling at air side or water side components (e.g., heating and cooling coil, air filter, cooling tower), (4) senso ...
+
+
+\section{Operational Faults Modeling}\label{operational-faults-modeling}
```
The list of FaultModel object types a little below this is incomplete.

rongpengzhang replied:
Addressed.

mjwitte commented on doc/engineering-reference/src/operational-faults.tex:
```diff
@@ -315,3 +318,134 @@ \subsubsection{Modeling Approach}\label{modeling-approach-3}
+
+\textbf{(a) constant flow chillers}
+
+For the chillers with emph{ConstantFlow} and emph{NotModulated}, local control is provided by resetting the leaving water temperature. The actual evaporator outlet water temperature value at faulty operations can be obtained via:
```
Two occurrences of emph need backslashes here.
![image](https://cloud.githubusercontent.com/assets/4991075/17002999/990985b8-4e93-11e6-86b5-07707fb69fbe.png)

rongpengzhang replied:
Addressed.

mjwitte commented on doc/engineering-reference/src/operational-faults.tex:
```diff
@@ -315,3 +318,134 @@ \subsubsection{Modeling Approach}\label{modeling-approach-3}
+
+\(T_{evap-i}\) evaporator inlet temperature value 
+
+\textbf{(b) variable flow chillers}
```
Should have some space before this heading, but not sure exactly how to make that happen.

rongpengzhang replied:
I find that"\newline" can solve this problem. Addressed.

mjwitte commented on src/EnergyPlus/FaultsManager.cc:
```diff
@@ -206,7 +265,7 @@ namespace FaultsManager {
 
 		// SUBROUTINE LOCAL VARIABLE DECLARATIONS:
-		static bool RunMeOnceFlag( false );
+		static bool RunFaultMgrOnceFlag( false );
```
@rongpengzhang I think we want this and other statics to be moved up into an anonymous namespace and then add a clear_state function to reset them?  @Myoldmopar @EnergyArchmage ?

rongpengzhang replied:
Addressed this for the FaultManager routine.

mjwitte commented on src/EnergyPlus/FaultsManager.cc:
```diff
@@ -218,394 +277,722 @@ namespace FaultsManager {
+						ErrorsFound = true;
+					} else {
+					// Link the chiller with the fault model
+						for ( int ChillerNum = 1; ChillerNum <= PlantChillers::NumElectricChillers; ++ChillerNum ) {
```
@rongpengzhang Why search the chiller list again here?  FindItemInList above has already searched for the matching chiller, so you could save that result as ChillerNum then use it in the if and use it directly here.  Similar for the other chillers.

rongpengzhang replied:
Yes, that's a much better and efficient design. Implemented it.

mjwitte commented on tst/EnergyPlus/unit/FaultsManager.unit.cc:
```diff
@@ -192,3 +200,66 @@ TEST_F( EnergyPlusFixture, FaultsManager_FaultFoulingAirFilters_CalFaultyFanAirF
 
 }
+
+TEST_F( EnergyPlusFixture, FaultsManager_FaultChillerSWTSensor_CalFaultChillerSWT )
```
@rongpengzhang These are good tests for CalFaultChillerSWT, but there is a lot of new code sprinkled among the various chiller models.  The new integration test covers only one chiller type Chilled:Electric:Reformulated:EIR.  Do you have integration test files for all of the other chiller types?  I am not suggesting that we add that many new test suite files, but we need to test all of this new code with something.

rongpengzhang replied:
Mike, this function is applicable for all the chiller types that are involved in this fault feature, not just Chiller:Electric:Reformulated:EIR. I checked it when I applied it to different chiller types.

mjwitte commented:
@Myoldmopar or @EnergyArchmage Would appreciate one of you making a pass through the code changes here.

mjwitte commented:
@rongpengzhang Tested this with the new Fault_ChillerSWTSensorOffset_RefBldgLargeOfficeNew2004.idf and compared to the same file without the fault.  The change in results look reasonable.  There's a typo at the end of the  new example file `!*** OPERAIONAL FAULTS ***` and the last line of the file needs a return.  I have no further comments at this point.

rongpengzhang committed:
Address final comments

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang commented:
@mjwitte Thanks so much for the careful review. I have addressed all your comments and pushed a new commit. I will notify Edwin to merge it if there is no CI warning for this commit.

EnergyArchmage commented on src/EnergyPlus/FaultsManager.cc:
```diff
@@ -206,406 +276,700 @@ namespace FaultsManager {
+		EvapOutletTemp = EvapOutletTemp_f;
+		EvapMassFlowRate = EvapMassFlowRate_f;
+		QEvaporator = QEvaporator_f;
+	}
```
I am sorry I am just looking at this feature for the first time.  This approach seems problematic to me.  There is no checking to see if the new evaporator outlet temperature is feasible or the new evaporator heat rate is within limits or capacity of the machine.  Just because a sensor goes haywire, you cannot assume the chiller can then always meet that setpoint.  

The reformulated EIR chiller uses the outlet temperature in the performance curves and I think there is a fundemental flaw with that in the approach implemented here.  For example you need to rerun the performance cureves with the new outlet temperature.  

Are these faults supposed to be inside the chiller machine or the supervisory control system's chilled water sensor downstream?  It would seem far more easy and robust for the faults to adjust the setpoints on the nodes and the let the chillers run unchanged, was that approach rejected for some reason?

rongpengzhang replied:
@EnergyArchmage Hi Brent, thanks for reviewing this feature. Please find my replies/clarifications to your three comments below:

(1) I agree with your concern about the reasonable level of evaporator outlet temperature at faulty cases. In the coding I enforce it to be within the range between ElecReformEIRChiller( EIRChillNum ).TempLowLimitEvapOut and Node( EvapInletNode ).Temp. The corresponding codes are here: ChillerReformulatedEIR.cc Line 2083

(2) The EvapOutletTempSetPoint is updated to the new (faulty) outlet temperature (Line 2083) before the chiller performance calculations based on curves (Line 2102). So, I don't think we need to rerun it.

(3) This fault feature is more complex than just to adjust the evaporator outlet setpoint. Let's consider these two cases: (a) a fault-free variable flow chiller with 5C setpoint, and (b) there is an faulty offset of 2C for a chiller with 7C setpoint. In (a), the chiller will determine water flow based on 5C and provide correct cooling amount that meets the required load; in (b), however, the chiller will determine water flow based on 7C although the actual evaporator outlet temperature is 5C. Therefore, the chiller will provide more cooling than required. I think the approach you suggested will behave as case (a) but what we need is case (b). That's why I put the fault code in the chiller calculations instead of the supervisory-level setpoint determinations.

Please let me know if you have further comments.

Just a general question about Github comment: is there a way to refer to a specific code line in our discussions?

EnergyArchmage replied:
okay, I am not sure your "b" case is very common, because variable flow is usually set by the coils not the chiller.  (except in old example files)

What seems odd to me is that if there is more than one chiller operating together but one supply temperature sensor that is faulty, the user has to enter fault input for each chiller.

I will pull down the branch and take a closer look at the chiller code.

EnergyArchmage replied:
I am not sure that the fault  will really make  "the chiller will provide more cooling than required" 

For example in the EIR chiller there is this code executed after the call to CalFaultChillerSWT

```
    // If load exceeds the distributed load set to the distributed load
    if ( QEvaporator > std::abs( MyLoad ) ) {
        if ( EvapMassFlowRate > MassFlowTolerance ) {
            QEvaporator = std::abs( MyLoad );
            EvapDeltaTemp = QEvaporator / EvapMassFlowRate / Cp;
            EvapOutletTemp = Node( EvapInletNode ).Temp - EvapDeltaTemp;
```

This will revert the load back to the requested load and undo the fault.

rongpengzhang replied:
These codes [e.g., ChillerReformulatedEIR.cc, Line 2276] should execute before calling to CalFaultChillerSWT [e.g., ChillerReformulatedEIR.cc, Line 2288], so the load (QEvaporator) will not be reverted back. It's a mistake in the  "ChillerElectricEIR.cc". I will double check this for all chiller models.
I checked the following key parameters at the end of CalcReformEIRChillerModel function and find the values are what we expect: (1) EvapMassFlowRate (2) EvapOutletTemp & EvapDeltaTemp (3) QEvaporator

Myoldmopar replied:
@EnergyArchmage Do you feel this is resolved?  I started looking at it to review it, and as I read along I was thinking the same thing.  It started to feel like the fault had the possibility of changing the physical state, not just the control logic.  I'll keep digging in here, but I was wondering your current thoughts as I do.

EnergyArchmage replied:
Yes, last time I looked that particular call did get moved to be later in ChillerElectricEIR.cc.  

I found that I needed pull up the code in the IDE because the github diff view was too narrow to see the context.  After studying that I got a little more comfortable with the approach.

EnergyArchmage commented on src/EnergyPlus/ChillerIndirectAbsorption.cc:
```diff
@@ -1559,6 +1564,19 @@ namespace ChillerIndirectAbsorption {
 		CpFluid = GetSpecificHeatGlycol( PlantLoop( IndirectAbsorber( ChillNum ).CWLoopNum ).FluidName, EvapInletTemp, PlantLoop( IndirectAbsorber( ChillNum ).CWLoopNum ).FluidIndex, RoutineName );
+		
+		//If there is a fault of Chiller SWT Sensor (zrp_Jun2016)
+		if( IndirectAbsorber( ChillNum ).FaultyChillerSWTFlag && ( ! WarmupFlag ) && ( ! DoingSizing ) && DoWeathSim ){
```
This use of DataGlobals::DoWeathSim in many places seems odd.  It isn't clear if you intend for the faults models to apply if the user sets Run Simulation for Sizing Periods to Yes.  (This is not for sizing, but for running the design days with the full model.)  With what you have here the behavior during sizing periods will change with the setting for Run Simulation for Weather File Run Periods.   Flags like DoWeathSim don't really inform what type of simulation is currently happening, for that it might be better to check DataGlobals::KindOfSim.

rongpengzhang replied:
Yes, you are right. I will add DataGlobals::DoDesDaySim in addition to DataGlobals::DoWeathSim.

EnergyArchmage replied:
I guess I was thinking DoWeathSim should either go away completely, or be replaced by something like 
       (  ( DataGloabals::KindOfSim == ksRunPeriodWeather ) || ( DataGloabals::KindOfSim == ksDesignDay ) )

It can also be good to exclude kick off timesteps using 

```
   ! DataGlobals::KickOffSimulation
```

rongpengzhang replied:
Got it.

rongpengzhang commented:
Comment:
I am not sure your "b" case is very common, because variable flow is usually set by the coils not the chiller. 
Reply:
For constant flow chillers, the flow rate is constant, so there should be no problem for this type of chiller.
For the variable flow chillers, the flow rate is calculated in the chiller model, not the coil side. For example
    EvapMassFlowRate = max( 0.0, ( QEvaporator / Cp / EvapDeltaTemp ) );  [ChillerReformulatedEIR.cc, Line 2185]
What the fault model does is to modify this value as well as the EvapOutletTemp in the chiller calculations [e.g., CalcReformEIRChillerModel in ChillerReformulatedEIR.cc].
We think the faults will lead to a situation that is similar to that in an over-sized or under-sized system. Since the current plant controls (e.g., bypass control) can well handle over-sized and under-sized situations, we think the faulty-sensor cases should also be well handled.

Comment:
What seems odd to me is that if there is more than one chiller operating together but one supply temperature sensor that is faulty, the user has to enter fault input for each chiller.
Reply:
In this situation, the user needs to create two fault objects, each corresponding to one chiller. I think this design is similar to that for the chiller setpoint: although two chillers may share the same leaving setpoint, user need to define two setpoint objects, each corresponding to one chiller.

rongpengzhang committed:
Address Brent's comments

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang committed:
Address build warnings

EnergyArchmage commented:
Okay, I have check over the latest code changes and they look good.

rongpengzhang commented:
Thanks, Brent!

@Myoldmopar Mike and Brent have reviewed this feature. Can you get this branch merged?

mjwitte commented:
@rongpengzhang Please bring this branch up to date one more time for a clean CI report, then I will merge.

rongpengzhang commented:
Thanks, @mjwitte. I think the codes are up to date. I notice that there are 4 examples failed in the CI tests, all of which are related with fault models. These diffs are generated because I added (!  DataGlobals::KickOffSimulation ) in the activation of all fault models. I have checked the absolute values of these diffs and find they are negligible.

mjwitte commented:
#5766 is up to date, but this one is still 14 commits behind when viewed from [here](https://github.com/NREL/EnergyPlus/branches/active)

rongpengzhang committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

rongpengzhang commented:
Got it. Just merged in the most updated Develop branch.

mjwitte commented on doc/engineering-reference/src/operational-faults.tex:
```diff
@@ -315,3 +318,139 @@ \subsubsection{Modeling Approach}\label{modeling-approach-3}
+
+\(T_{evap-i}\) evaporator inlet temperature value 
+
+\newline \newline
```
@rongpengzhang The engineering ref does not build correctly on my machine.  It makes a pdf but says it failed with some errors like this "LaTeX error : There's no line here to end."  I suspect it doesn't like these \newline commands.  I would just go ahead and delete them here and below.

rongpengzhang replied:
@mjwitte 
According to https://www.sharelatex.com/learn/Line_breaks_and_blank_spaces, any of the following three ways should work for creating new lines:
(1) \ \ (two backslashes)
(2) \newline
(3) \hfill \break

I tested the fault section with "\newline" (not the whole engineering ref) on sharelatex.com and it works well. Maybe two \newline cannot be put together?

rongpengzhang replied:
@mjwitte I've replaced "\newline" with "\leavevmode \". I think this can solve the problem.

I've never built the whole engineering ref locally. So can you please test it for me? Very appreciate it.

Myoldmopar replied:
Wait.  Why are you forcing new lines here?  Just leave it and let LaTeX glue it all together.  If you want to polish it up a bit, add the list of variables to an itemized list, such as:

```
\end{equation}

Where,
\begin{itemize}
  \item \(m_{evap}\) evaporator water flow rate (design value, actual value)
  \item blahbladeeblah
\end{itemize}
```

Then LaTeX will do a good job of adding spaces after the itemized list.

mjwitte replied:
@Myoldmopar Without anything there's no space at all.  With \item, won't it put bullets in front?  Feel free to modify as you see fit.

Myoldmopar replied:
Yeah, though I'm OK with bullets. I'll compile it tomorrow and take a look
at what's going on.

rongpengzhang committed:
Debug LaTex Warning

mjwitte committed:
One more latex newline tweak

mjwitte commented:
@rongpengzhang The leavemodes left too much white space.  So, after some experimenting, found a placement for \newline that works.

![image](https://cloud.githubusercontent.com/assets/4991075/17235436/aa3f0b5e-5506-11e6-985b-c18126743a5f.png)

Myoldmopar committed:
Merge develop; minor doc conflict resolved

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rongpengzhang @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

Myoldmopar committed:
Fix branch idd object changes to get test file to pass; merge develop

Myoldmopar committed:
Update fault file to v8.6

Myoldmopar commented:
Dummy comment to figure out image things:

120694307_Fault_ChillerSWTSensorOffset (Myoldmopar) - x86_64-MacOS-10.9-clang: OK (2342 of 2346 tests passed, 566 test warnings)

 * 570 tests had: AUD diffs.
 * 4 tests had: EIO diffs.
 * 4 tests had: ESO big diffs.
 * 4 tests had: MTR big diffs.
 * 1 test had: Table small diffs.
 * 3 tests had: Table big diffs.

[![Build Badge](http://img.shields.io/badge/build%20status-passing-green.svg)](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)
[![Test Badge](http://img.shields.io/badge/tests%20passed-99.83%25-yellow.svg)](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)

lefticus commented:
Test image posting.

120694307_Fault_ChillerSWTSensorOffset (Myoldmopar) - x86_64-MacOS-10.9-clang: OK (2342 of 2346 tests passed, 566 test warnings)

 * 570 tests had: AUD diffs.
 * 4 tests had: EIO diffs.
 * 4 tests had: ESO big diffs.
 * 4 tests had: MTR big diffs.
 * 1 test had: Table small diffs.
 * 3 tests had: Table big diffs.


<a href='https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.svg)</a> <a href='https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-99.83%25-yellow.svg)</a>

lefticus commented:
Dummy comment to figure out image things:

120694307_Fault_ChillerSWTSensorOffset (Myoldmopar) - x86_64-MacOS-10.9-clang: OK (2342 of 2346 tests passed, 566 test warnings)

 * 570 tests had: AUD diffs.
 * 4 tests had: EIO diffs.
 * 4 tests had: ESO big diffs.
 * 4 tests had: MTR big diffs.
 * 1 test had: Table small diffs.
 * 3 tests had: Table big diffs.

[![Build Badge](http://img.shields.io/badge/build%20status-passing-green.svg)](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)
[![Test Badge](http://img.shields.io/badge/tests%20passed-99.83%25-yellow.svg)](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)

Myoldmopar commented:
ANOTHER Dummy comment to figure out image things:

120694307_Fault_ChillerSWTSensorOffset (Myoldmopar) - x86_64-MacOS-10.9-clang: OK (2342 of 2346 tests passed, 566 test warnings)

 * 570 tests had: AUD diffs.
 * 4 tests had: EIO diffs.
 * 4 tests had: ESO big diffs.
 * 4 tests had: MTR big diffs.
 * 1 test had: Table small diffs.
 * 3 tests had: Table big diffs.

![image](http://img.shields.io/badge/build%20status-passing-green.svg)[Link](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)

![image](http://img.shields.io/badge/tests%20passed-99.83%25-yellow.svg)[Link](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html)

Myoldmopar commented:
I have a good feeling about this test...

120694307_Fault_ChillerSWTSensorOffset (Myoldmopar) - x86_64-MacOS-10.9-clang: OK (2342 of 2346 tests passed, 566 test warnings)

 * 570 tests had: AUD diffs.
 * 4 tests had: EIO diffs.
 * 4 tests had: ESO big diffs.
 * 4 tests had: MTR big diffs.
 * 1 test had: Table small diffs.
 * 3 tests had: Table big diffs.


<a href='https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-39db73a045bd6a60715b4f3701644048fb862118-x86_64-MacOS-10.9-clang.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-99.83%25-yellow.png)</a>

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 120694307_Fault_ChillerSWTSensorOffset

Myoldmopar committed:
Add destructors in fault manager class structure

Myoldmopar commented:
The only failure was the custom check license year thing, and I don't think this adds any new license text, so it should merge just fine.  Merging this.  Finally.

rongpengzhang commented:
Happy to see this branch is finally got merged. Thanks for all the review and comments @mjwitte @Myoldmopar Happy New Year!
